[UPDATE] Plugin: git

ncanceill opened the pull request:
## Update branch

**Plugin:** `git`

This branch contains updates for the `git` plugin. If you want to add a commit, please [submit a PR](https://github.com/ncanceill/oh-my-zsh/pull/new/plugin-git) to [my branch](https://github.com/ncanceill/oh-my-zsh/tree/plugin-git).

## Status

**READY FOR MERGE**

This is now rebased against latest upstream (ff4663a6b83b8dd205da3968de4bc458abc08b2c).

There is still some stuff to do before merge:
- [x] Test completion now that most `compdef`s were removed
- [x] Settle the debate on `gp/gl` versus `gP/gp`
- [x] Do another sweep to include any pending PRs related to this plugin
- [ ] Update the [wiki](https://github.com/robbyrussell/oh-my-zsh/wiki/Plugin:git)
- [x] Squash commits, because atomicity is out of the window already

## Changelog

The auto-closing feature [should kick in](#issuecomment-43131152) when this is merged, which will ease maintenance.

* Fixes #1500
* Fixes #1504
* Fixes #1546
* Fixes #1601
* Fixes #1800
* Fixes #1966
* Fixes #1982
* Fixes #2124
* Fixes #2338
* Fixes #2407
* Fixes #2433
* Fixes #2463
* Fixes #2493
* Uses #2494, fixes #1846
* Fixes #2546
* Fixes #2684
* Fixes #2714
* Fixes #2921
* Fixes #3085
* Fixes #3091
* Fixes #3122
* Fixes #3276
* Fixes #3374
* Fixes #3543
* Fixes #3680, fixes #3708
* Fixes #3711
* Fixes #3726
* Fixes #3749
* Fixes #3763
* Fixes #3836

mcornella commented:
I stand in favor of pushing this PR. A lot of work has been put into this and it would provide an immense benefit in terms of closed issues and feature requests asked for since long ago. Whether we want a simpler 

@ncanceill you have my support to solve current merge conflicts and implement the pending issues. I've been running these changes for months and I've built a workflow around it.

Let's get to work!

robbyrussell commented:
Echoing what @mcornella said.... @ncanceill -- if you could rebase and send my way, I'll get this merged.

Kriechi commented:
@ncanceill I think this PR fixes #3543 as well.

Stibbons commented:
What about splinting this file into several plugins so one can only enable the aliases he really wants, and avoid conflicts on tools on his system (see https://github.com/robbyrussell/oh-my-zsh/issues/3682)

ncanceill commented:
Hi guys,

Finally rebased and sorted aliases. Yay!

 I believe I did not miss any of the updates that got to upstream in the meantime; in case I did, please notify ASAP. I did, however, not include modifications conflicting with changes previously present in this PR.

I included all the changes previously present in this PR. I sorted all aliases alphabetically (a process during which I deleted a few duplicates), I removed the unnecessary `compdef` statements and the call to `xargs` as suggested by @blueyed, and added some of the aliases suggested here and there in this thread.

There are still a few things to do before this can be merged. Please see the updated message at the top.

@blueyed, can you confirm that I correctly removed the unnecessary `compdef`s? Also, I too find some of those "potentially destructive aliases" pretty useful, can you live with them?

@Stibbons, it is your choice whether you want to be listed as maintainer. Concerning #3682, I do not think spreading the aliases over several files would solve the problem, since there would typically be no more than a few conflicting aliases anyway; I believe the idea of a warning when overriding an alias would work better.

@mcornella, would you help me browse through the open PRs to see if there anything more to include here?

@robbyrussell, thanks for the kind words, I appreciate your trust, and will notify you when this is ready for merge.

**To all:** can we get some testing/advice/remarks/suggestions? can we, like we did on merge vs rebase when pulling (see #2071), finally settle on `gp/gl` vs `gP/gp`?

Happy Easter/Passover/egg-hunt/whatever-you-celebrate to all!

Stibbons commented on plugins/git/git.plugin.zsh:
```diff
@@ -1,177 +1,188 @@
+alias gke='\gitk --all $(git log -g --pretty=format:%h)'
+compdef _git gke='gitk'
+
+alias gl='git pull'
```
For me this is confusing. gl is pull while glp is git log...
And you don't have gpr I use everyday (git pull --rebase)

I still think gp = git pull, gP = git push, gl = git log, is more logical.

ncanceill replied:
We would probably need a kind of vote to decide this.

Stibbons commented on plugins/git/git.plugin.zsh:
```diff
@@ -1,177 +1,188 @@
+alias gunignore='git update-index --no-assume-unchanged'
+alias gunwip='git log -n 1 | grep -q -c "\-\-wip\-\-" && git reset HEAD~1'
+alias gup='git pull --rebase'
+compdef _git gup=git-fetch
```
are you sure of this compdel?

ncanceill replied:
Well, `git-pul` and `git-fetch` both work, so I will change it (and the one below for `gupv`) for the sake of consistency.

Stibbons commented on plugins/git/git.plugin.zsh:
```diff
@@ -1,177 +1,188 @@
+compdef _git gpoat=git-push
+alias gpu='git push upstream'
+
+alias gr='git remote'
```
given that you have "grba"/"grbc", I would recommend to rename all these gr1, gr2... to grbN (grb = git rebase), grm = git remote.

ncanceill replied:
This may satisfy a need for consistency, but I am not sure it really helps, and it would make those aliases one character longer, so I do not think this is the right choice.

Stibbons commented:
ok for me if you remove my name from maintainer.

I have given a set of remarks. For me, on the overall it is pretty good, while I would just make one or two more step toward alias unification...

For the conflict detector, this should be a method done in the lib folder that would check if an command already exist and avoid creating the alias (and maybe print a message?). I cannot look at this for the moment unfortunately

ncanceill commented:
Thanks for your remarks @Stibbons. I will get started on an alias conflict detector.

Stibbons commented:
+1

ncanceill commented:
OK, commits are squashed and properly rebased, I think we are good to go.

@robbyrussell merge at will, I will update the wiki when you do.

blueyed commented on plugins/git/git.plugin.zsh:
```diff
@@ -1,177 +1,212 @@
-    git log --pretty=$1
-  fi
-}
+alias gl='git pull'
```
This should be "git log ...".

ncanceill replied:
We already discussed `gl/gp` vs `gp/gP`.

blueyed commented on plugins/git/git.plugin.zsh:
```diff
@@ -1,177 +1,212 @@
+alias glol="git log --graph --pretty=format:'%Cred%h%Creset -%C(yellow)%d%Creset %s %Cgreen(%cr) %C(bold blue)<%an>%Creset' --abbrev-commit"
+alias glola="git log --graph --pretty=format:'%Cred%h%Creset -%C(yellow)%d%Creset %s %Cgreen(%cr) %C(bold blue)<%an>%Creset' --abbrev-commit --all"
+alias glog='git log --oneline --decorate --color --graph'
+alias glone='git clone'
```
For `git clone` `gcl` seems to be better.

I currently use this function for it:

    gcl () {
      git clone --recursive $@ || return
      local last=$@[$#]
      if [[ -d $last ]]
      then
        cd $last
      else
        local dir=${${last:t}%.git}
        if [[ -d $dir ]]
        then
          cd $dir
        fi
      fi
    }

When using just an alias, it should have `--recursive`, too.

ncanceill replied:
Done, `git config --list` will now be `gcf`.

blueyed commented on plugins/git/git.plugin.zsh:
```diff
@@ -1,177 +1,212 @@
+
+alias gp='git push'
+alias gpd='git push --dry-run'
+alias gpf='git push --force'
```
This should not be an alias. It's destructive and `gp -f` can be used instead.

ncanceill replied:
Done.

blueyed commented on plugins/git/git.plugin.zsh:
```diff
@@ -1,177 +1,212 @@
+alias gr7='git rebase -i HEAD~7'
+alias gr8='git rebase -i HEAD~8'
+alias gr9='git rebase -i HEAD~9'
+alias gr10='git rebase -i HEAD~10'
```
If these are added, they should be named `grbX`, but I really think that it's not worth it.

Stibbons replied:
agree

ncanceill replied:
Can I remove them altogether? It is not long to write `grbi @~2`.

blueyed replied:
:+1:

ncanceill replied:
Done.

blueyed commented on plugins/git/git.plugin.zsh:
```diff
@@ -1,177 +1,212 @@
+alias gstaa='git stash apply'
+alias gstd='git stash drop'
+alias gstl='git stash list'
+alias gstp='git stash pop'
```
I am using this function for this currently:

    gstp () {
      local log stash_ref
      if [[ "$1" = <-> ]] || [[ -z "$1" ]]
      then
        stash_ref="stash@{${1:-0}}" 
      else
        stash_ref="$1" 
      fi
      log=$(git log -g --pretty="%s" $stash_ref -1)  || {
        echo "Failed to display log for $stash_ref."
        return 1
      }
      local current_branch="$(current_branch)"
      if ! [[ $log = "WIP on "${current_branch}* ]]
      then
        echo "Warning: stash appears to be for another branch."
        echo "Current: ${current_branch}"
        echo "Log: $log"
        echo -n "Continue? "
        read -q || {
          echo
          return
        }
        echo
      fi
      git stash pop $stash_ref
    }

ncanceill replied:
I like it, but I am not sure it should replace `git stash pop`, maybe stand as a verbose alternative? `gstv`?

blueyed replied:
True, and I would not add it to oh-my-zsh, given how many lines it adds.

blueyed commented on plugins/git/git.plugin.zsh:
```diff
@@ -1,177 +1,212 @@
+alias gstl='git stash list'
+alias gstp='git stash pop'
+alias gsts='git stash show --text'
+alias gsu='git submodule update'
```
I find `gsmu` better for this - using `gsm` as prefix for `git submodule`.

ncanceill replied:
We already have `gsi` for another submodule alias, and I think the idea of prefixes went out the window a long time ago anyway.

blueyed commented on plugins/git/git.plugin.zsh:
```diff
@@ -1,177 +1,212 @@
-alias gignored='git ls-files -v | grep "^[[:lower:]]"'
+alias gunwip='git log -n 1 | grep -q -c "\-\-wip\-\-" && git reset HEAD~1'
+alias gup='git pull --rebase'
+alias gupv='git pull -v --rebase'
```
Move the `-v` to the end here, to make it better match the previous line.

ncanceill replied:
Done. You are very picky, are you not?

Stibbons commented:
I have a remark: this PL takes too much time, is there any way for someone to get merge permission on  the plugins of oh-my-zsh?

natalia-mentormate commented:
:+1:  and I agree with @Stibbons that it will be great if someone can have it merged..

ncanceill commented:
Are we good here?

mcornella commented:
@ncanceill quick question, what is the `git_cmd` stuff at the top of the file for? There is `command git` to ensure we don't use an alias or a function, but the zstyle stuff has me confused...

Will merge this one soon so it's urgent, thank you!

blueyed commented:
> what is the git_cmd stuff at the top of the file for

It will use the same `git` as zsh's vcs_info does.

But I think that using `command git` as a default might be better.  But that could also get done after this PR is merged.

mcornella commented:
Ok thanks @blueyed. You're right though, it's more important to put this out there than changing this little thing. Thanks!

robbyrussell commented:
Wow. You folks are awesome. This was a long-time coming!

mcornella commented:
Thank you @ncanceill! :heart:

yous commented:
Nice work! I love it. :+1:

fbiville commented:
Why, oh why, have there been aliases removed?
I have been using `ggpush` and `ggpur` everyday for 3 years, what am I supposed to do here? Define the aliases myself?

So much for backwards compatibility :( :-1:

JunkieLand commented:
Same for me ! Those aliases were not conflicting with other expected definitions. Why remove them ?

yous commented:
@fbiville @JustinAiken It seems that [`ggp`](https://github.com/robbyrussell/oh-my-zsh/pull/2790/files#diff-4ad2773a26526d0e361e66139c283099R107) replaces `ggpush`, [`ggu`](https://github.com/robbyrussell/oh-my-zsh/pull/2790/files#diff-4ad2773a26526d0e361e66139c283099R117) replaces `ggpur`. Now they become shorter, but we should remember again. :sweat_smile:

blueyed commented:
@fbiville @JunkieLand 
Why not check the diff?  It's `ggp` now.. see https://github.com/robbyrussell/oh-my-zsh/pull/2790/files#diff-4ad2773a26526d0e361e66139c283099R107.

mcornella commented:
I agree with you guys, I executed it poorly. I'm currently working on [updating the wiki entries](https://github.com/robbyrussell/oh-my-zsh/wiki/Plugin%3Agit#aliases) but in the meantime @yous is right: `ggp` is `ggpush`, `ggl` is `ggpull` and `ggu` is `ggpur`.

Sorry!

fbiville commented:
@blueyed @yous good to hear they are not completely gone.
@mcornella why not keeping the old aliases, just for the sake of backwards compat?

razor-x commented:
Perhaps when removing long-standing aliases like this, instead of just having them randomly disappear without warning, one could add a deprecation notice. Something like:

Before (good):

```zsh
$ gap
...works normally...
```

Now (bad)

```zsh
$ gap
zsh: command not found: gap
```

What it should be (better)

```zsh
$ gap
The gap alias will be removed in the next major update.
...works normally...
```

Max101 commented:
@mcornella :-1:  ggpull, ggpur, ggpush removed? gosh people kind of get used to aliases, I have been using this for more than a year now... Would be cool to keep those...

yachi commented:
It took a year and tons of effort by a bunch of people to get this massive rewrite done and get merged, which is not easy already. Everyone could join the process. Please appreciate the effort. The rename process may be not elegant enough, but you can always define your own alias!

And Yay this is finally merged 🎉

hypn commented:
Agreed, why remove "ggpull" etc commands. You realize many of us type these commands many times a day, and have been for years? Some kind of deprecation message (eg: echo "Please us ggl from now " in the alias) would have gone a long way. Not happy :-1:

devunt commented:
And these commits also caused #3962.

Schabernack commented:
the git add --patch alias seems to have been removed without a replacement. was that intentional?

ncanceill commented:
Hi all,

Thank you @BlindHunter for all the hate, and much thanks to everybody else for all the love!

@robbyrussell, I and all who contributed simply wanted to improve the awesome framework you created =D

***

Before addressing what may have gone wrong, I believe the priority is to solve the newly introduced issues:
* [ ]  `local` is used outside a function — #3962
* [ ] the `gg*` aliases were renamed with shorter names — #3972
* [ ] the `gap` alias was removed — #3972
* [ ] the `gdc` alias conflicts with GNU D compiler — #3977

Please notify me ASAP if there are more.

***

### Proposed fixes

#### `loca

devunt commented:
@ncanceill Nevermind. You did a (almost) great job. Contributing to an open source project needs a lot of passion and it should be welcomed unless they did it in bad purpose.

Let's don't blame him/her.

BlindHunter commented:
Sorry :) I believe this could be filtered on the code review stage.

ncanceill commented:
@BlindHunter that is OK, just remember to be cordial when asking for help or reporting issues, nobody pays me for this.

Our updating process is extremely poor right now, and all of this just shows that it drastically needs to be improved. You are welcome to join the discussion, which will start as soon as we fix the issues mentioned above.

Once again, thank you all for your support, and do not forget to thank all the other who contributed to this PR.

Douglas-Su commented:
`git_cmd` was confused me. each time I run `source ~/.zshrc`, `git_cmd=git` will pop up which makes me unconformable...

heliostatic commented:
Another related but slightly different issue is that the current_branch function has changed. Whereas the previous function,
```
function current_branch() {
ref=$(git symbolic-ref HEAD 2> /dev/null) || \
ref=$(git rev-parse --short HEAD 2> /dev/null) || return
echo ${ref#refs/heads/}
}

CorvinM commented:
@ncanceill  Why not check for existence of the program before adding each alias so that any other installed program takes priority over the git aliases?

ncanceill commented:
@Douglas-Su what you are experiencing is due to the `local` call at plugin initialization, and should be fixed shortly

***

@heliostatic thanks for reporting, can you elaborate on what the `current_branch` function does on your computer?

***

@CorvinM yes, we are working on an aliasing system which would take care of such things

heliostatic commented:
@ncanceill It appears to do nothing? Running it in a git repo returns nothing, as does running outside a git repo.

KenPowers commented:
`gwip` has some strange output for me now:

![image](https://cloud.githubusercontent.com/assets/174864/8098088/1ef91a28-0fb9-11e5-9c3c-e3dc7019a721.png)

robbyrussell commented:
Sorry, everyone. We'll get this sorted out soon enough

ncanceill commented:
@heliostatic this is weird, can you check what happens when you run these four commands successively in a Git repo?
```zsh
git symbolic-ref --quiet HEAD
echo $?
git rev-parse --short HEAD
echo $?
```
Also could you post your Git config? ([gist](https://gist.github.com/) it if it is too long to paste here).

***

@KenPowers thanks for reporting, this is due to the new form of the alias and will be fixed shortly; indeed, when your WIP commit does not contain deleted files, you will see this error message — however the command should still work despite the error message, does it not?

heliostatic commented:
@ncanceill Thanks:
```

KenPowers commented:
@ncanceill Yes, it still works. It just looks funny.

heliostatic commented:
@ncanceill I'm guessing my issue is related to the `local` keyword issue that you identified in #2790

ncanceill commented:
@KenPowers thanks for confirming, this will be fixed ASAP

@heliostatic in this case, would you be so kind as to fetch my fork and cherry-pick 6c29041 to see if it fixes your issue? also could you try without all those aliases in your Git config? (probably unrelated but just to be sure).

ncanceill commented:
OK @KenPowers would you be so kind as to fetch my fork and cherry-pick 6ff96da to confirm that it fixes the issue with `gwip`?

heliostatic commented:
@ncanceill fetched and cherry-picked https://github.com/robbyrussell/oh-my-zsh/commit/6c29041af73fc3668f6a3256ebc7ab532a2bbba9 which did fix my issue. Thanks!

ncanceill commented:
@heliostatic awesome! I will note this as fixed, thank you for testing.

Mezzle commented:
Just as a small point.  I used to be able to do `ggpush --force` - I can't do `ggp --force`, as whilst the alias expanded, the function doesn't.

ncanceill commented:
@Mezzle yes this has been reported and will be fixed shortly, in the meantime you can use `ggf` as short for `ggp --force`

maya commented:
What's the status of `git_cmd=git` popping up at the top of the new shell?

ncanceill commented:
@maya this is fixed in #3990 which should be merged very soon

mblayman commented:
@ncanceill Thanks for taking on the role of maintainer! I'm sure that is not easy work. I noticed that the `gaa` alias to `git add --all` was removed as well. Could that be re-included? I find it useful because I work on projects with deep tree structures so `ga .` will not always do what I want.

mcornella commented:
@mblayman the `gaa` alias will be included again in #3990. Wait for it...

mcgeeco commented:
When I ```echo $(current_branch)``` I get the current branch, and I when I run ```git pull origin $(current_branch)``` it works as expected, but when I run ggl for some reason it seems to call ```git pull origin HEAD```.

ncanceill commented:
That is weird, the `ggl` function specifically uses `current_branch` when called with no argument.

Can you check that you are running the proper `ggl` function by using `which ggl`?

mcgeeco commented:
@ncanceill 
![screenshot from 2015-06-17 09 31 00](https://cloud.githubusercontent.com/assets/5881581/8202629/a23e2094-14d3-11e5-8d91-426a26f10afc.png)

ncanceill commented:
OK, I thought I got rid of those empty string problems, but I was wrong.

```zsh
git pull origin "$(current_branch)"

From github.com:ncanceill/test
 * branch            master     -> FETCH_HEAD
Already up-to-date.

git pull origin "$(current_branch)" ""

From github.com:ncanceill/test
 * branch            master     -> FETCH_HEAD
 * branch            HEAD       -> FETCH_HEAD
Already up-to-date.
```

I will send a PR with the fix today, please standby.

mcgeeco commented:
@ncanceill Great, thanks! Let me know if I can help with anything.

chaucerbao commented:
How are you guys remembering the new `ggu` and `ggp` commands?  The letters **u** and **p** are in both words.  I know `ggpur` and `ggpush` are back in the mix, but I'm trying to use the new, 40-50% more efficient and super productive aliases.

dogus-akaydin commented on plugins/git/git.plugin.zsh:
```diff
@@ -1,177 +1,202 @@
-alias gd='git diff'
-gdv() { git diff -w "$@" | view - }
-compdef _git gdv=git-diff
-alias gdt='git difftool'
```
Just curious, why is there not a replacement alias for `git difftool`?
